airbyte-integrations/connectors/source-microsoft-onedrive/source_microsoft_onedrive/spec.py

```diff
@@ -0,0 +1,110 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+
+from typing import Any, Dict, Literal, Union
+
+import dpath.util
+from airbyte_cdk.sources.file_based.config.abstract_file_based_spec import AbstractFileBasedSpec
+from pydantic import BaseModel, Field
+
+
+class OAuthCredentials(BaseModel):
+    """
+    OAuthCredentials class to hold authentication details for Microsoft OAuth authentication.
+    This class uses pydantic for data validation and settings management.
+    """
+
+    class Config:
+        title = "Authenticate via Microsoft (OAuth)"
+
+    # Fields for the OAuth authentication, including tenant_id, client_id, client_secret, and refresh_token
+    auth_type: Literal["Client"] = Field("Client", const=True)
+    tenant_id: str = Field(title="Tenant ID", description="Directory (tenant) ID")
+    client_id: str = Field(
+        title="Client ID",
+        description="Client ID of your Microsoft developer application",
+        airbyte_secret=True,
+    )
+    client_secret: str = Field(
+        title="Client Secret",
+        description="Client Secret of your Microsoft developer application",
+        airbyte_secret=True,
+    )
+    refresh_token: str = Field(
+        title="Refresh Token",
+        description="Refresh Token of your Microsoft developer application",
+        airbyte_secret=True,
+    )
+
+
+class ServiceCredentials(BaseModel):
+    """
+    ServiceCredentials class for service key authentication.
+    This class is structured similarly to OAuthCredentials but for a different authentication method.
+    """
+
+    class Config:
+        title = "Service Key Authentication"
+
+    # Fields for the Service authentication, similar to OAuthCredentials
+    auth_type: Literal["Service"] = Field("Service", const=True)
+    tenant_id: str = Field(title="Tenant ID", description="Directory (tenant) ID")
+    client_id: str = Field(
+        title="Client ID",
+        description="Client ID of your Microsoft developer application",
+        airbyte_secret=True,
+    )
+    client_secret: str = Field(
+        title="Client Secret",
+        description="Client Secret of your Microsoft developer application",
+        airbyte_secret=True,
+    )
+
+
+class SourceMicrosoftOnedriveSpec(AbstractFileBasedSpec, BaseModel):
```

Comment:
@lazebnyi you might still work on this, but this is currently going through all the drives the user has access to which could be an enormous amount of files - a strategy like used in other filed based connectors like S3 or Drive where the user has to pick a folder/bucket to work with is probably necessary in practice.

Reply:
@flash1293 Thanks for comment. But in that way we restrict users and they should create connection for each bucket/folder. wdyt?
May be better move this field to stream configuration level form main level as a compromise?

Reply:
It would also work on the stream level, no super strong opinion here. The thing we need to avoid is that listing files goes through thousands of pages without giving the user a way to control it.

Reply:
Added drive name filed to main level

Reply:
I still think we need to go to the folder level (at least for each stream) - for example the shared company drive for Airbyte (Google Drive, but similar concept) contains many thousand documents.

Reply:
Added